lib/native/source/opensmt/api.patch

```diff
@@ -0,0 +1,537 @@
+     PTRef dividend = args[0];
+     PTRef divisor = args[1];
+-    if (not isConstant(divisor) and not isConstant(dividend)) { throw LANonLinearException("Divisor or dividend must be constant in linear logic"); }
++    if (not isConstant(divisor)) { throw LANonLinearException("Divisor or dividend must be constant in linear logic"); }
+     if (isZero(divisor)) { throw ArithDivisionByZeroException(); }
+     if (isOne(divisor)) { return dividend; }
+     if (isMinusOne(divisor)) { return mkNeg(dividend); }
+@@ -1177,7 +1177,7 @@ std::string ArithLogic::printTerm_(PTRef tr, bool ext, bool safe) const
+         char* tmp_str;
+         opensmt::stringToRational(tmp_str, sym_store.getName(getPterm(tr).symb()));
+         opensmt::Number v(tmp_str);
+-        if (!isNonNegNumConst(tr)) {
++        if (!isNonNegative(v)) {
+             v.negate();
+             is_neg = true;
+         }
+diff --git a/src/logics/ArithLogic.h b/src/logics/ArithLogic.h
+index 6f2669f9..fa8a7e26 100644
+--- a/src/logics/ArithLogic.h
++++ b/src/logics/ArithLogic.h
+@@ -4,7 +4,7 @@
+ #include "Number.h"
+ #include <numeric>
+ 
+-class LANonLinearException : std::runtime_error
++class LANonLinearException : public std::runtime_error
+ {
+     std::string msg;
+ public:
+@@ -417,4 +417,4 @@ public:
+     bool operator ()  (PTRef x_, PTRef y_) const;
+ 
+ };
+-#endif
+\ No newline at end of file
++#endif
+diff --git a/src/logics/FunctionTools.h b/src/logics/FunctionTools.h
+index fc9dbfb7..943a466d 100644
+--- a/src/logics/FunctionTools.h
++++ b/src/logics/FunctionTools.h
+@@ -33,7 +33,7 @@ class TemplateFunction {
+     FunctionSignature signature;
+     PTRef tr_body;
+ 
+-    inline static std::string_view constexpr template_arg_prefix = ".arg";
++    inline static constexpr std::string_view template_arg_prefix = ".arg";
+     inline static std::size_t template_arg_counter = 0;
+ public:
+     static std::string nextFreeArgumentName() { return std::string(template_arg_prefix) + std::to_string(template_arg_counter++); }
+diff --git a/src/logics/Logic.cc b/src/logics/Logic.cc
+index b90430c2..1161b068 100644
+--- a/src/logics/Logic.cc
++++ b/src/logics/Logic.cc
+@@ -15,6 +15,7 @@
+ #include "Substitutor.h"
+ #include "smt2tokens.h"
+ #include "IteHandler.h"
++#include "Interpret.h"
+ 
+ #include <iostream>
+ #include <map>
+@@ -32,7 +33,7 @@ const char* Logic::e_argnum_mismatch = "incorrect number of arguments";
+ const char* Logic::e_bad_constant    = "incorrect constant for logic";
+ 
+ const char* Logic::tk_val_uf_default   = "UFDefault";
+-const char* Logic::tk_val_bool_default = "true";
++const char* Logic::tk_val_bool_default = "false";
+ 
+ const char* Logic::tk_true     = "true";
+ const char* Logic::tk_false    = "false";
+@@ -130,12 +131,13 @@ bool Logic::hasQuotableChars(std::string const & name) const
+         0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+         0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+         0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
++        0, 0, 0, 0, 0, 0
```

Comment:
Why do we need to patch a number of zeroes?

Reply:
Did they just fill an array with 256 chars with only 250 instanzes of 1 or 0? Oh no ;-)

Reply:
Yes, the array is too short :)
I ran into this while debugging one of the test, and while it's not the actual bug, this might also have caused problems for us since JavaSMT does allow all 256 ASCII codes.

Reply:
I took a look at the OpenSMT code. This specific method seems to be quite weird. It seems to suffer from several issues:
- the array is too short (fixed by adding more zeros)
- the array-access is not guaranteed to be in range [0, 256] (fixed by casting to unsigned char)
- the large 0/1-block is not readable and might require some documentation.
- C++ might not store strings as char-arrays (8bit), but as int32-arrays, such that additional prefix-bits from UTF-8 appear in the comparison (fixed by casting to unsigned char).
- Does C/C++ even guarantee that a char is always 8bit wide? It is the default, but the standard might allow arbitrary sizes for characters and bytes. (Just a meta question about machine-independent code style :-) )

The benefit is that it might be quite fast, if checking millions of characters.

One could replace that whole code with something much simpler:
```
if (name.front() == '|' and name.back() == '|') return false; // Already quoted
for (auto c : name) {
   switch (c) { // check for quotable chars
       case ' ':
       case '"':
       case '#':
       case '$':
       case '\'':
       case ':':
       case ';':
       case '[':
       case '\\':
       case ']':
       case '{':
       case '}':
           return true;
   }
}
return false;
```

This might avoid several low-level C/C++-based issues, and it is surely more readable than the previous code.

@daniel-raffler Could you file a pull-request for that change in OpenSMT?